### DISCUSSION TITLE
最新模版与之前模版vscode编译设置区别

### DISCUSSION THREAD

**Discussion by qiaoxiaoyue** [sjtug/SJTUThesis#693] [提问] [answered]
我19年时在MAC上使用VScode使用当时的模版进行了论文撰写。最近下载了新的模版，依然使用之前VScode中的编译设置，显示Latexmk: Initialization file '.latexmkrc' gave an error:
     Undefined subroutine &main::ensure_path called at (eval 12) line 10, <GEN0> chunk 1.

Latexmk: Stopping because of problem with rc file
想知道当前模版与早期模版在VScode中设置的区别，不想重新配置，怎样可以同时兼容两个版本的模版？

**Comment by skyzh** [comment]
可能是 TeXLive 版本太低？升级到最新的 TeXLive 2021 试试？

**Reply by qiaoxiaoyue** [reply]
升级了版本还是不行

**Reply by qiaoxiaoyue** [reply]
我错了，只升级了编辑器，忘记升级TeXLive

**Reply by Dawntown** [reply]
之前我是TeX Live 2017, 也遇到了一样的问题, 然后升级到2021以后又提示biblatex的版本太高了, 这可如何是好... 能给一个requirements吗?
```
SyncTeX written on main.synctex.gz.Latexmk: Non-existent bbl file 'main.bbl in line'
 No file main.bbl.
Latexmk: Run number 1 of rule 'biber main'
Biber error: [123] Utils.pm:185> ERROR - Error: Found biblatex control file version 3.7, expected version 3.4.
```

**Reply by skyzh** [reply]
可以清理一下文件：latexmk -C

**Reply by Dawntown** [reply]
> 可以清理一下文件：latexmk -C

没用, 用`latexmk -C`清理完中间文件还是一样的报错

**Reply by skyzh** [reply]
可能是卸载/安装不完全导致的，参考 https://tex.stackexchange.com/questions/294531/texstudio-biber-claims-biblatex-version-is-wrong

**Reply by Dawntown** [reply]
确实是没有卸载干净之前的texlive 2017.
仔细看报错应该是biber和biblatex版本不兼容导致的, 使用`which biber`发现在`/usr/local/texlive/2021/bin`以外的`/usr/bin`里也有biber命令, 并且还有好多有关latex的其他命令, 在执行了以下命令以后再编译就成功了
```bash
sudo apt remove texlive # 会显示出之前没有卸载干净的包, 我的texlive 2017是通过apt安装的, 因此这样卸载
sudo apt autoremove
```

**Comment by AlexaraWu** [comment, marked as answer]
目前的配置文件 `.latexmkrc` 需要 Latexmk 版本 4.59 或以上。

**Reply by qiaoxiaoyue** [reply]
升级了版本之后，之前的文件无法编译，请问有可以兼容的配置嘛？或者怎么做可以同时编译新老两个版本呢？

**Reply by skyzh** [reply]
可以把具体的报错拿出来新开一个 discussion。有个包在 TeXLive 2021 里面和之前不兼容了。
